### PR TITLE
Fixed Voice Activity Command crashing on high ages

### LINKED ISSUE #300: VC Activity command crashes
**Describe the bug**
The vc activity command crashes when the max-age exceeds about 1 million.

**Screenshots**
image

--- 

This could be fixed by, for example, implementing a max value for the max-age.

### PR DESCRIPTION
**Pull request by RealYusufIsmail**
fixed the issue were when age option was set to a huge number caused the command to crash. The max is 365 days and the min is 1 day now. 

This closes #300

### CODE REVIEW

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java [hunk 1]
```diff
@@ -86,7 +86,8 @@
             new OptionData(OptionType.STRING, MAX_USES_OPTION,
                     "The amount of times the invite can be used, default is infinity", false),
             new OptionData(OptionType.INTEGER, MAX_AGE_OPTION,
-                    "Max age in seconds. Set this to 0 to never expire, default is 1 day", false));
+                    "Max age in seconds. Set this to 0 to never expire, default is 1 day", false)
+                        .setRequiredRange(1, 365));
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java:90]: 1 to 365 seconds is definitely not correct.

**Review by Zabuzard — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java (hunk 1 above)
Comment (RealYusufIsmail) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java]: Changed to 604800 and that is correct since it says in the guild.

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java [hunk 2]
```diff
@@ -84,9 +84,11 @@
 
     private static final List<OptionData> inviteOptions = List.of(
             new OptionData(OptionType.STRING, MAX_USES_OPTION,
-                    "The amount of times the invite can be used, default is infinity", false),
+                    "The amount of times the invite can be used, default is infinity", false)
+                        .setRequiredRange(0, 100),
             new OptionData(OptionType.INTEGER, MAX_AGE_OPTION,
-                    "Max age in seconds. Set this to 0 to never expire, default is 1 day", false));
+                    "Max age in seconds. Set this to 0 to never expire, default is 1 day", false)
+                        .setRequiredRange(0, 604800));
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java:91]: This number is an unreadable mess. Use the `java.time` API to make it clear what this is.

**Review by Zabuzard — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java (hunk 2 above)
Comment (RealYusufIsmail) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java]: Changed it to use java.time.Duration

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java [hunk 3]
```diff
@@ -121,10 +126,10 @@ public VcActivityCommand() {
     public void onSlashCommand(@NotNull SlashCommandEvent event) {
         Member member = Objects.requireNonNull(event.getMember(), "member is null");
         GuildVoiceState voiceState = Objects.requireNonNull(member.getVoiceState(),
-                "Voicestates aren't being cached, check the JDABuilder");
+                "Voice states aren't being cached, check the JDABuilder");
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java:129]: Not for this PR and disagree

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java [hunk 4]
```diff
@@ -121,10 +126,10 @@ public VcActivityCommand() {
...
 
         if (!voiceState.inVoiceChannel()) {
-            event.reply("You need to be in a voicechannel to run this command!")
+            event.reply("You need to be in a voice channel to run this command!")
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java:132]: Not for this PR

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java [hunk 5]
```diff
@@ -37,6 +38,8 @@
 public final class VcActivityCommand extends SlashCommandAdapter {
     private static final Logger logger = LoggerFactory.getLogger(VcActivityCommand.class);
 
+    private static final Duration DURATION = Duration.ofDays(7);
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java:41]: What duration?

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java [hunk 6]
```diff
@@ -84,9 +87,11 @@
 
     private static final List<OptionData> inviteOptions = List.of(
             new OptionData(OptionType.STRING, MAX_USES_OPTION,
-                    "The amount of times the invite can be used, default is infinity", false),
+                    "The amount of times the invite can be used, default is infinity", false)
+                        .setRequiredRange(0, 100),
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java:91]: Use constants instead of magic numbers

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java [hunk 7]
```diff
@@ -84,9 +87,11 @@
...
             new OptionData(OptionType.INTEGER, MAX_AGE_OPTION,
-                    "Max age in seconds. Set this to 0 to never expire, default is 1 day", false));
+                    "Max age in seconds. Set this to 0 to never expire, default is 1 day", false)
+                        .setRequiredRange(0, DURATION.getSeconds()));
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java:94]: Use constants instead of magic numbers

**Review by Tais993 — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java (hunk 5 above)
Comment (RealYusufIsmail) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java]: Changed to MAX_USE_DURATION

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java (hunk 6 above)
Comment (RealYusufIsmail) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java]: Changed to MIN_USE_AMOUNT and MAX_USE_AMOUNT

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java (hunk 7 above)
Comment (RealYusufIsmail) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java]: Changed to MIN_AGE_DURATION and MAX_AGE_DURATION

**Review by Tais993 — APPROVED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java [hunk 8]
```diff
@@ -95,7 +95,7 @@
                         .setRequiredRange(MIN_USE_DURATION, MAX_USE_DURATION),
             new OptionData(OptionType.INTEGER, MAX_AGE_OPTION,
                     "Max age in seconds. Set this to 0 to never expire, default is 1 day", false)
-                        .setRequiredRange(MIN_AGE_DURATION, MAX_AGE_DURATION.getSeconds()));
+                        .setRequiredRange(MIN_AGE_DURATION.getSeconds(), MAX_AGE_DURATION.getSeconds()));
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java]: Just do `0, TimeUnit.DAYS.toSeconds(7)` instead.

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java [hunk 9]
```diff
@@ -38,7 +38,7 @@
 public final class VcActivityCommand extends SlashCommandAdapter {
     private static final Logger logger = LoggerFactory.getLogger(VcActivityCommand.class);
 
-    private static final int MIN_AGE_DURATION = 0;
+    private static final Duration MIN_AGE_DURATION = Duration.ofSeconds(0);
     private static final Duration MAX_AGE_DURATION = Duration.ofDays(7);
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java]: Remove those obsolete fields-

**Review by Zabuzard — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java (hunk 9 above)
Comment (illuminator3) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java]: Tijs told him to add them.

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java (hunk 3 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java]: Imo mini fixes like typos can be done in another PR as well. Other than that I would agree though.

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java (hunk 9 above)
Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java]: ^^^
To keep consistency, and against magic numbers.

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java (hunk 8 above)
Comment (RealYusufIsmail) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java]: Changed it now.

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java [hunk 10]
```diff
@@ -37,6 +38,12 @@
 public final class VcActivityCommand extends SlashCommandAdapter {
     private static final Logger logger = LoggerFactory.getLogger(VcActivityCommand.class);
 
+    private static final int MIN_AGE_DURATION = 0;
+    private static final long MAX_AGE_DURATION = TimeUnit.DAYS.toSeconds(7);
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java:42]: The constants should have the unit in their name to prevent confusion. `MIN/MAX_AGE_DURATION_SECONDS`

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java [hunk 11]
```diff
@@ -84,9 +91,11 @@
 
     private static final List<OptionData> inviteOptions = List.of(
             new OptionData(OptionType.STRING, MAX_USES_OPTION,
-                    "The amount of times the invite can be used, default is infinity", false),
+                    "The amount of times the invite can be used, default is infinity", false)
+                        .setRequiredRange(MIN_USE_AMOUNT, MAX_USE_AMOUNT),
             new OptionData(OptionType.INTEGER, MAX_AGE_OPTION,
-                    "Max age in seconds. Set this to 0 to never expire, default is 1 day", false));
+                    "Max age in seconds. Set this to 0 to never expire, default is 1 day", false)
+                        .setRequiredRange(MIN_AGE_DURATION, MAX_AGE_DURATION));
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java:98]: Since you are using min and max now, is the `handleIntegerTypeOption` still needed? If not, remove. (@Tais993 please confirm, thanks)

**Review by Zabuzard — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java (hunk 11 above)
Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java:98]: The logic checking or it's a valid Integer isn't needed, but the null checks and such still are.

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java:98]: Okay, then lets just adjust it a bit and all cool, I guess.

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java (hunk 10 above)
Comment (RealYusufIsmail) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java]: Fixed

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java [hunk 12]
```diff
@@ -38,8 +38,8 @@
 public final class VcActivityCommand extends SlashCommandAdapter {
     private static final Logger logger = LoggerFactory.getLogger(VcActivityCommand.class);
 
-    private static final int MIN_AGE_DURATION = 0;
-    private static final long MAX_AGE_DURATION = TimeUnit.DAYS.toSeconds(7);
+    private static final int  MIN_AGE_DURATION_SECONDS = 0;
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java]: You have two spaces now `int__MIN`, also shows me that you pushed without applying spotless - dont.

**Review by Zabuzard — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java (hunk 12 above)
Comment (RealYusufIsmail) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java]: Fixed

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java]: Its not. You changed completely different parts of the code.

**Review by Zabuzard — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java (hunk 12 above)
Comment (RealYusufIsmail) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java]: I ran spotless. I will change it manually

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java [hunk 13]
```diff
@@ -157,22 +153,22 @@ public void onSlashCommand(@NotNull SlashCommandEvent event) {
...
+        OptionalInt maxAge;
 
         // the user already received the error in the handleIntegerTypeOption method
         // it still throws to tell us to return this method and stop the proceeding code
         try {
-            maxAge = handleIntegerTypeOption(event, maxAgeOption);
+            maxAge = getIntegerOptionIfPresent(maxAgeOption);
         } catch (IllegalArgumentException ignore) {
             return;
         }
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java:174]: The try-catch is completely obsolete, this exception can not be thrown here...

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java [hunk 14]
```diff
@@ -191,7 +187,8 @@ public void onSlashCommand(@NotNull SlashCommandEvent event) {
                     getKeyByValue(VC_APPLICATION_TO_ID, applicationId).orElse("an activity");
         }
 
-        handleSubcommand(event, voiceChannel, applicationId, maxUses, maxAge, applicationName);
+        handleSubcommand(event, voiceChannel, applicationId, maxUses.getAsInt(), maxAge.getAsInt(),
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java:190]: And what if the optional is not present? Then this `.getAsInt()` call crashes! You clearly havent tested this. This is a critical bug.

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java [hunk 15]
```diff
@@ -235,48 +232,18 @@ private static void handleErrors(@NotNull SlashCommandEvent event,
...
      * <p>
      * <p/>
      *
-     * @param event the {@link SlashCommandEvent}
-     * @param optionMapping the {@link OptionMapping}
-     * @return nullable {@link Integer}
-     * @throws java.lang.IllegalArgumentException if the option's value is - outside of
-     *         {@link Integer#MAX_VALUE} - negative
+     * @param optionMapping the {@link OptionMapping} to validate
      */
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java:241]: The Javadoc is in a bad state right now, the text is hard to understand and not fitting. Also, what is this `<p><p/>` mess? And why do you take in the `event`, it is not even used!

Rewrite:
```java
* Interprets the option as integer, if present.
*
* @param option the option to interpret or null if not present
* @return the integer contained in the option, if present
```

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java [hunk 16]
```diff
@@ -235,48 +232,18 @@ private static void handleErrors(@NotNull SlashCommandEvent event,
...
-                .setEphemeral(true)
-                .queue();
-            throw new IllegalArgumentException(optionMapping.getName() + " can't be negative");
-        }
-
-
-        return optionValue;
+    private static OptionalInt getIntegerOptionIfPresent(@Nullable OptionMapping optionMapping) {
+        return Optional.ofNullable(optionMapping)
+            .stream()
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java:244]: NIT: I dislike the `.stream()` hack. Its a dirty abuse of the API. I know its very common on StackOverflow though... your choice.

```java
return optionMapping == null
  ? OptionalInt.empty();
  : OptionalInt.of(Math.toIntExact(optionMapping.getAsLong()));
```

Also, gets rid of the unecessary boxing conversion from `long` to `Long` that `map(OptionMapping::getAsLong)` introduces...

**Review by Zabuzard — CHANGES_REQUESTED** [review]

**Review by Tais993 — CHANGES_REQUESTED** [review]
As I explained in the original PR, and in the discord last night.

Using optional is useless, null is a valid value, using null makes it a ton easier.

If we set the max-age to null, it'll use the default value.
Otherwise we've to add an if check and such, which just add code without any purpose

**Comment by Zabuzard** [comment]
> As I explained in the original PR, and in the discord last night.
> 
> Using optional is useless, null is a valid value, using null makes it a ton easier.
> 
> If we set the max-age to null, it'll use the default value. Otherwise we've to add an if check and such, which just add code without any purpose

Are you talking about the bigger context, i.e. where the `OptionalInt` will be used again later? I.e. the setter where its passed to?

Not sure why you think you have to add ifs, you can just write `setFoo(value.orElse(null))`. But sure, I havent thought about the use-site yet, only about the creation-site.

That said, I could argue that methods should be designed in an isolated fashion and not with how their call-site will use them in mind - and then `OptionalInt` is cleaner. But its a private method anyways, so its only purpose is to assist the surrounding code as best as it can, making `@Nullable Integer` totally valid again.

**tl;dr**: I dont care, I accept both variants now.

**Comment by Tais993** [comment]
You can just use orElse yeah, but adds code, which I don't like since it doesn't improve anything.

I'd go for Integer, the author of this PR may decide.

**Comment by RealYusufIsmail** [comment]
Will continue next week I have had a tests this week including one mock and 2 normal test tomorrow

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java [hunk 17]
```diff
@@ -187,8 +163,9 @@ public void onSlashCommand(@NotNull SlashCommandEvent event) {
                     getKeyByValue(VC_APPLICATION_TO_ID, applicationId).orElse("an activity");
         }
 
-        handleSubcommand(event, voiceChannel, applicationId, maxUses.getAsInt(), maxAge.getAsInt(),
-                applicationName);
+        handleSubcommand(event, voiceChannel, applicationId,
+                event.getOption(MAX_USES_OPTION).getAsLong(),
+                event.getOption(MAX_AGE_OPTION).getAsLong(), applicationName);
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java]: This code doesn't even work, besides that, still extremely disagree to usage of Optional.

Optional exists to "wrap" around null, so handling becomes "easier" (subjective)
So why for gods sake, use Optional if it's literally a value that we encourage. (in this context)

This means you need to use `optional.orElse(null)` which honestly, is really damn useless 
Adds code, for no reason. Not for readability, maintainability, just, because you can?

Completely disagree with that.

Also talked about why we shouldn't use Optional on your [other PR](https://github.com/Together-Java/TJ-Bot/pull/303#pullrequestreview-824980517)

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java [hunk 18]
```diff
@@ -204,12 +181,12 @@ public void onSlashCommand(@NotNull SlashCommandEvent event) {
...
             @NotNull VoiceChannel voiceChannel, @NotNull String applicationId,
-            @Nullable Integer maxUses, @Nullable Integer maxAge, @NotNull String applicationName) {
+            @Nullable Long maxUses, @Nullable Long maxAge, @NotNull String applicationName) {
 
         voiceChannel.createInvite()
             .setTargetApplication(applicationId)
-            .setMaxUses(maxUses)
-            .setMaxAge(maxAge)
+            .setMaxUses(Math.toIntExact(maxUses))
+            .setMaxAge(Math.toIntExact(maxAge))
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java]: Why use a long if we want an int?
Currently the flow is like:

Int -> long -> int

Why not just keep it an int?

**Review by Tais993 — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java [hunk 19]
```diff
@@ -182,7 +163,16 @@ public void onSlashCommand(@NotNull SlashCommandEvent event) {
...
-        handleSubcommand(event, voiceChannel, applicationId, maxUses, maxAge, applicationName);
+        OptionMapping maxUsesOption = event.getOption(MAX_USES_OPTION);
+        OptionMapping maxAgeOption = event.getOption(MAX_AGE_OPTION);
+
+        if (maxUsesOption == null || maxAgeOption == null) {
+            handleSubcommand(event, voiceChannel, applicationId, 0, 86400, applicationName);
+        } else {
+            handleSubcommand(event, voiceChannel, applicationId,
+                    getOptionAsInteger(maxUsesOption).getAsInt(),
+                    getOptionAsInteger(maxAgeOption).getAsInt(), applicationName);
```
Comment (krankkkk) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java:174]: - The If doesn't really make sense, if just one of the 2 is null you also ignore the other
 -  Whats the point of the OptionalInt? In else they will always exist.
 - Why not squash the if/else into an `getOptionalAsInteger(...).orElse(defaultValue)`?

Comment (krankkkk) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/VcActivityCommand.java:174]: Also, magic numbers what does 0 / 86400 mean in this context?

**Comment by Zabuzard** [comment]
Ownership of this is transferred.

**Comment by Zabuzard** [comment]
The current state of this PR is not very helpful, its probably better to do it from scratch again with all the learnings earned in the discussions here. Closing for now.
